README.md
Crypto Dip Buying Bot
=====================
This bot is designed to buy cryptocurrency on Coinbase Pro or Gemini using a USD prefunded portfolio whenever it detects a significant dip in price.

USE AT YOUR OWN RISK
--------------------
I run this bot full time against my own personal Coinbase Pro and Gemini accounts, however I make no warranties that
the bot will function. It could crash and miss a dip, or it could detect and buy a dip before the floor. So far
it has done well for me, but your mileage may vary.  
As with any open source code: **USE THIS BOT AT YOUR OWN RISK!**

Dip Detection
-------------
The bot checks the price in a configurable cycle. Each cycle the bot will check the price of the specified cryptocurrency.
It will then compare the average price of the previous 7 days worth of price history to the configured dip percentage.
If the current price is the configured percentage lower than the price average it will buy the cryptocurrency in the
specified amount of USD.

Running The Bot
---------------
To run the bot you will need Docker and docker-compose installed on your computer.  

    docker-compose up -d

Choosing An Exchange
--------------------
If you specify Gemini credentials at all in the `config.json` file then the bot will use Gemini even if Coinbase Pro
credentials are also specified.

Config File
-----------
You will need the following:

1. Coinbase Pro or Gemini credentials tied to the portfolio you want to run the bot against
2. Dip logic parameters:
    1. The cryptocurrency you want to transact in. (It must support being paired against USD in Coinbase Pro)
    2. The buy amount you want in $USD.
    3. The average percentage drop from the previous week's worth of intervals you want to consider a buy worthy dip.

The following sections are optional.

1. Time variables in the bot config
   1. Period of days to average (Default: 7)
   2. Cool down period before buying again (Default: 7)
   3. Check cycle frequency in minutes (Default: 60)
2. AWS credentials:
   1. AWS API keys
   2. SNS topic ARN (us-east-1 only for now)
3. Optionally you can override the bot name

These settings should be in a configuration file named `config.json` and placed in `./config`.
Additionally, you can override the volume mount to a new path if you prefer.
The file should look like this:

```json
{
  "bot": {
    "currency": "ETH",
    "buy_amount": 75.00,
    "dip_percentage": 10,
     "average_period_days": 3,
     "cool_down_period_days": 5,
     "cycle_time_minutes": 15,
     "name": "Test-Bot"
  },
  "coinbase": {
    "api_key": "YOUR_API_KEY",
    "api_secret": "YOUR_API_SECRET",
    "passphrase": "YOUR_API_PASSPHRASE"
  },
   "gemini": {
    "api_key": "YOUR_API_KEY",
    "api_secret": "YOUR_API_SECRET",
   },
   "aws": {
    "access_key": "YOUR_API_KEY",
    "secret_access_key": "YOUR_API_SECRET",
    "sns_arn": "arn:aws:sns:us-east-1:012345678901:dip_alerts"
  }
}
```

Running outside of Docker
-------------------------
You can run the bot outside of Docker pretty easily.

```bash
python SourceCode/cryptodip-bot.py -c /path/to/config.json
```

Logs
----
The bot will log activity to stdout, so you can review it with `docker logs`

Donations
---------
Any and all donations are greatly appreciated.  
I have GitHub Sponsors configured however I happily prefer cryptocurrency:

ETH/ERC20s: ahrenstein.eth (0x288f3d3df1c719176f0f6e5549c2a3928d27d1c1)  
BTC: 3HrVPPwTmPG8LKBt84jbQrVjeqDbM1KyEb


CHANGELOG.md
Crypto Dip Buying Bot: Changelog
================================
A list of all the changes made to this repo, and the bot it contains

Version 0.3.1-r1
----------------
This update only changes the repository CI/CD process and documentation

1. Update [TESTING.md](TESTING.md) to reflect proper information
2. Update [GitHub Actions](/.github/workflows) to deploy to DockerHub now that they don't permit auto-building for free


Version 0.3.1
-------------

1. Minor fixes around Gemini API calls when Gemini is having a service outage.
The bot should no longer crash but API outages do prevent data from being gathered or Txs occurring

Version 0.3.0
-------------

1. Shifted TODO versions around to reflect new goals.
2. **BREAKING CHANGE** Added Gemini Support
    1. The database name the bot uses is now exchange-currency-"bot", so a new DB will be created when
    using the new version of the bot.
    2. Additionally, you have the option to give the bot a custom name so you can run more than one bot against the same
    exchange/currency pair.
3. Fixed a bug where price data would not continue gathering if the bot was not funded.
4. Super basic exception catching around DB functions.
5. `test-compose.yml` for local debugging/testing is now separate from the production example `docker-compose-yml`
6. Added the ability to override the hourly cycle to a different cycle specified in minutes.
7. Fixed a mistake in the default starting config

Version 0.2.0
-------------

1. **Fixed a critical issue where the last buy date is not updated when buys are made!**
2. Bot functions condensed in to 1 file
3. Averaging time period and purchase cool down period are now both config variables
4. Fixed some incorrect function descriptions

Version 0.1.0
-------------

1. Initial Pre-release of repository

Return to [README](README.md)


SourceCode/mongo.py
#!/usr/bin/env python3
"""Functions to use with MongoDB"""
#
# Python Script:: mongo.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#

import datetime
import pymongo
import bot_internals

# Constants that might be useful to adjust for debugging purposes
PURGE_OLDER_THAN_DAYS = 30


def add_price(bot_name: str, db_server: str, current_price: float):
    """Add a current price record to the database

    Args:
    bot_name: The name of the bot
    db_server: The MongoDB server to connect to
    current_price: The current price of the currency
    """
    timestamp = datetime.datetime.utcnow()
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        prices_collection = bot_db["prices"]
        record = {"time": timestamp, "price": current_price}
        prices_collection.insert_one(record)
    except Exception as err:
        print("Error creating price record: %s" % err)


def read_all_prices(bot_name: str, db_server: str):
    """Read all current price records in the database

    Args:
    bot_name: The name of the bot
    db_server: The MongoDB server to connect to
    """
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        prices_collection = bot_db["prices"]
        records = prices_collection.find()
    except Exception as err:
        print("Error reading price records: %s" % err)
    for record in records:
        print(record)


def average_pricing(bot_name: str, db_server: str, average_period: int) -> float:
    """Check the last week of prices and return the average

        Args:
        bot_name: The name of the bot
        db_server: The MongoDB server to connect to
        average_period: The time period in days to average across

        Returns:
        average_price: The average price of the last week
        """
    price_history = []
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        prices_collection = bot_db["prices"]
        records = prices_collection.find({})
    except Exception as err:
        print("Error reading price records for averaging: %s" % err)
    for record in records:
        record_age = datetime.datetime.utcnow() - record['time']
        if record_age.days <= average_period:
            price_history.append(record['price'])
    average_price = bot_internals.get_average(price_history)
    return average_price


def cleanup_old_records(bot_name: str, db_server: str):
    """Remove all price history older than X days

    Args:
    bot_name: The name of the bot
    db_server: The MongoDB server to connect to
    """
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        prices_collection = bot_db["prices"]
        records = prices_collection.find()
    except Exception as err:
        print("Error cleaning up old price records: %s" % err)
    for record in records:
        record_age = datetime.datetime.utcnow() - record['time']
        if record_age.days >= PURGE_OLDER_THAN_DAYS:
            prices_collection.delete_one({"_id": record['_id']})


def set_last_buy_date(bot_name: str, db_server: str):
    """Sets the date the last time the currency was bought

    Args:
    bot_name: The name of the bot
    db_server: The MongoDB server to connect to
    """
    timestamp = datetime.datetime.utcnow()
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        buy_date = bot_db["buy-date"]
    except Exception as err:
        print("Error connecting to buy-date collection: %s" % err)
    try:
        buy_date.find_one_and_update({"_id": 1},
                                     {"$set": {"time": timestamp}}, upsert=True)
    except Exception as err:
        print("Error updating buy date record: %s" % err)


def check_last_buy_date(bot_name: str, db_server: str,
                        cool_down_period: int) -> bool:
    """Get the date of the last time the currency was bought
    and returns true if it >= cool down period

    Args:
    bot_name: The name of the bot
    db_server: The MongoDB server to connect to
    cool_down_period: The time period in days that you will wait before transacting

    Returns:
    clear_to_buy: A bool that is true if we are clear to buy
    """
    try:
        # Create a Mongo client to connect to
        mongo_client = pymongo.MongoClient(db_server)
        bot_db = mongo_client[bot_name]
        buy_date = bot_db["buy-date"]
    except Exception as err:
        print("Error connecting to buy-date collection: %s" % err)
    # Create an initial record if the record doesn't exist yet
    if buy_date.find({'_id': 1}).count() == 0:
        print("Initializing new last buy date")
        timestamp = datetime.datetime.utcnow()
        buy_date.find_one_and_update({"_id": 1},
                                     {"$set": {"time": timestamp}}, upsert=True)
        return False
    try:
        last_buy_date = buy_date.find_one({"_id": 1})['time']
    except Exception as err:
        print("Error getting buy date record: %s" % err)
        return False
    time_difference = datetime.datetime.utcnow() - last_buy_date
    return time_difference.days >= cool_down_period


SourceCode/bot_math.py
#!/usr/bin/env python3
"""Simple math functions"""
#
# Python Script:: bot_math.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#


def get_average(list_of_numbers: list) -> float:
    """Read all current price records in the database

    Args:
    list_of_numbers: A list of floats

    Returns:
    list_average: A float containing the list average
    """
    sum_of_numbers = 0
    for number in list_of_numbers:
        sum_of_numbers = sum_of_numbers + number
    list_average = sum_of_numbers / len(list_of_numbers)
    return round(list_average, 2)


def dip_percent_value(price: float, percent: float) -> float:
    """Return the value of the current price if it dips a certain percent

    Args:
    price: The price to check a dip percentage against
    percent: the dip percentage we care about

    Returns:
    dip_price: A float containing the price if we hit our dip target
    """
    dip_price = price * (1 - percent / 100)
    return round(dip_price, 2)


SourceCode/bot_alerts.py
#!/usr/bin/env python3
"""Alerting Functions"""
#
# Python Script:: bot_alerts.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#


import json
import boto3


def get_aws_creds_from_file(config_file: str) -> [str, str, str]:
    """Open a JSON file and get AWS credentials out of it
    Args:
        config_file: Path to the JSON file containing credentials

    Returns:
        aws_access_key: The AWS access key your bot will use
        aws_secret_key: The AWS secret access key
        sns_topic_arn: The SNS topic ARN to publish to
    """
    with open(config_file) as creds_file:
        data = json.load(creds_file)
    aws_access_key = data['aws']['access_key']
    aws_secret_key = data['aws']['secret_access_key']
    sns_topic_arn = data['aws']['sns_arn']
    return aws_access_key, aws_secret_key, sns_topic_arn


def post_to_sns(aws_access_key: str, aws_secret_key: str, sns_topic_arn: str,
                message_subject: str, message_body: str):
    """Read all current price records in the database

    Args:
    aws_access_key: The AWS access key your bot will use
    aws_secret_key: The AWS secret access key
    sns_topic_arn: The SNS topic ARN to publish to
    message_subject: A message subject to post to SNS
    message_body: A message body to post to SNS
    """
    sns = boto3.client('sns', region_name="us-east-1",
                       aws_access_key_id=aws_access_key, aws_secret_access_key=aws_secret_key)
    sns.publish(TopicArn=sns_topic_arn, Subject=message_subject, Message=message_body)


SourceCode/coinbase_pro.py
#!/usr/bin/env python3
"""Functions to use with Coinbase Pro"""
#
# Python Script:: coinbase_pro.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#

import base64
import json
import time
import hmac
import hashlib
import requests
from requests.auth import AuthBase


# Create custom authentication for CoinbasePro
# as per https://docs.pro.coinbase.com/?python#creating-a-request
class CoinbaseProAuth(AuthBase):
    """
        Coinbase Pro provided authentication method with minor fixes
        """
    def __init__(self, api_key, secret_key, passphrase):
        self.api_key = api_key
        self.secret_key = secret_key
        self.passphrase = passphrase

    def __call__(self, request):
        timestamp = str(time.time())
        try:
            message = timestamp + request.method + request.path_url + (request.body or b'').decode()
        except:
            message = timestamp + request.method + request.path_url + (request.body or b'')
        hmac_key = base64.b64decode(self.secret_key)
        signature = hmac.new(hmac_key, message.encode(), hashlib.sha256)
        signature_b64 = base64.b64encode(signature.digest()).decode()

        request.headers.update({
            'CB-ACCESS-SIGN': signature_b64,
            'CB-ACCESS-TIMESTAMP': timestamp,
            'CB-ACCESS-KEY': self.api_key,
            'CB-ACCESS-PASSPHRASE': self.passphrase,
            'Content-Type': 'application/json'
        })
        return request


def get_cbpro_creds_from_file(config_file: str) -> [str, str, str]:
    """Open a JSON file and get Coinbase Pro credentials out of it
    Args:
        config_file: Path to the JSON file containing credentials and config options

    Returns:
        cbpro_api_key: An API key for Coinbase Pro
        cbpro_api_secret: An API secret for Coinbase Pro
        cbpro_api_passphrase: An API passphrase for Coinbase Pro
    """
    with open(config_file) as creds_file:
        data = json.load(creds_file)
    cbpro_api_key = data['coinbase']['api_key']
    cbpro_api_secret = data['coinbase']['api_secret']
    cbpro_api_passphrase = data['coinbase']['passphrase']
    return cbpro_api_key, cbpro_api_secret, cbpro_api_passphrase


def get_coin_price(api_url: str, config_file: str, currency: str) -> float:
    """
    Get the USD price of a coin from Coinbase Pro

    Args:
        api_url: The API URL for Coinbase Pro
        config_file: Path to the JSON file containing credentials and config options
        currency: The cryptocurrency the bot is monitoring

    Returns:
        coin_price: The price the coin currently holds in USD
    """
    # Instantiate Coinbase API and query the price
    coinbase_creds = get_cbpro_creds_from_file(config_file)
    coinbase_auth = CoinbaseProAuth(coinbase_creds[0], coinbase_creds[1], coinbase_creds[2])
    api_query = "products/%s-USD/ticker" % currency
    result = requests.get(api_url + api_query, auth=coinbase_auth)
    coin_price = float(result.json()['price'])
    return coin_price


def verify_balance(api_url: str, config_file: str, buy_amount: float) -> bool:
    """Check if enough money exists in the account
    Args:
        api_url: The API URL for Coinbase Pro
        config_file: Path to the JSON file containing credentials and config options
        buy_amount: The amount of $USD the bot plans to spend

    Returns:
        all_clear: A bool that returns true if there is enough money to transact
    """
    # Instantiate Coinbase API and query the price
    coinbase_creds = get_cbpro_creds_from_file(config_file)
    coinbase_auth = CoinbaseProAuth(coinbase_creds[0], coinbase_creds[1], coinbase_creds[2])
    api_query = "accounts"
    result = requests.get(api_url + api_query, auth=coinbase_auth).json()
    try:
        for account in result:
            if account['currency'] == "USD":
                if float(account['balance']) >= buy_amount:
                    return True
    except Exception as err:
        print("ERROR: Unable to get current balance!")
        print(err)
        return False
    # Return false by default
    return False


def buy_currency(api_url: str, config_file: str, currency: str, buy_amount: float) -> bool:
    """
    Conduct a trade on Coinbase Pro to trade a currency with USD

    Args:
        api_url: The API URL for Coinbase Pro
        config_file: Path to the JSON file containing credentials and config options
        currency: The cryptocurrency the bot is monitoring
        buy_amount: The amount of $USD the bot plans to spend

    Returns:
        trade_success: A bool that is true if the trade succeeded
    """
    coinbase_creds = get_cbpro_creds_from_file(config_file)
    # Instantiate Coinbase API and query the price
    coinbase_auth = CoinbaseProAuth(coinbase_creds[0], coinbase_creds[1], coinbase_creds[2])
    buy_query = 'orders'
    order_config = json.dumps({'type': 'market', 'funds': buy_amount,
                               'side': 'buy', 'product_id': '%s-USD' % currency})
    buy_result = requests.post(api_url + buy_query, data=order_config, auth=coinbase_auth).json()
    if 'message' in buy_result:
        print("LOG: Buy order failed.")
        print("LOG: Reason: %s" % buy_result['message'])
        return False
    else:
        print("LOG: Buy order succeeded.")
        print("LOG: Buy Results: %s" % json.dumps(buy_result, indent=2))
        return True


SourceCode/cryptodip_bot.py
#!/usr/bin/env python3
"""A bot that attempts to buy cryptocurrency on the dip."""
#
# Python Script:: cryptodip_bot.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#

import argparse
import bot_internals


def main(config_file: str, debug_mode: bool):
    """
    The main function that triggers and runs the bot functions

    Args:
    config_file: Path to the JSON file containing credentials and config options
    debug_mode: Use Sandbox APIs instead of production
    """
    # Load the configuration file
    config_params = bot_internals.read_bot_config(config_file)
    # Start the correct cycle
    if config_params[6]:
        bot_internals.gemini_exchange_cycle(config_file, debug_mode)
    else:
        bot_internals.coinbase_pro_cycle(config_file, debug_mode)


if __name__ == '__main__':
    # This function parses and return arguments passed in
    # Assign description to the help doc
    PARSER = argparse.ArgumentParser(
        description='A bot that attempts to buy'
                    ' cryptocurrency on the dip.')
    # Add arguments
    PARSER.add_argument(
        '-c', '--configFile', type=str, help="Path to config.json file", required=True
    )
    PARSER.add_argument(
        '-d', '--debug', help="Use sandbox APIs", required=False, action='store_true'
    )
    # Array for all arguments passed to script
    ARGS = PARSER.parse_args()
    ARG_CONFIG = ARGS.configFile
    ARG_DEBUG = ARGS.debug
    main(ARG_CONFIG, ARG_DEBUG)


SourceCode/bot_internals.py
#!/usr/bin/env python3
"""Internal functions the bot uses"""
#
# Python Script:: bot_internals.py
#
# Linter:: pylint
#
# Copyright 2021, Matthew Ahrenstein, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matt@ahrenstein.com
#
# See LICENSE
#

from itertools import count
import json
import datetime
import time
import boto3
import coinbase_pro
import gemini_exchange
import mongo


def read_bot_config(config_file: str) -> [str, float, int, int, int, bool, bool, int, str]:
    """Open a JSON file and get the bot configuration
    Args:
        config_file: Path to the JSON file containing credentials and config options

    Returns:
        crypto_currency: The cryptocurrency that will be monitored
        buy_amount: The price in $USD that will be purchased when a dip is detected
        dip_percentage: The percentage of the average price drop that means a dip occurred
        average_period_days: The time period in days to average across
        cool_down_period_days: The time period in days that you will wait before transacting
        aws_loaded: A bool to determine if AWS configuration options exist
        using_gemini: A bool to determine if the bot should use Gemini
        cycle_time_minutes: The cycle interval in minutes
        bot-name: The name of the bot
    """
    with open(config_file) as creds_file:
        data = json.load(creds_file)
    crypto_currency = data['bot']['currency']
    buy_amount = data['bot']['buy_amount']
    dip_percentage = data['bot']['dip_percentage']
    aws_loaded = bool('aws' in data)
    using_gemini = bool('gemini' in data)
    if 'average_period_days' in data['bot']:
        average_period_days = data['bot']['average_period_days']
    else:
        average_period_days = 7
    if 'cool_down_period_days' in data['bot']:
        cool_down_period_days = data['bot']['cool_down_period_days']
    else:
        cool_down_period_days = 7
    if 'cycle_time_minutes' in data['bot']:
        cycle_time_minutes = data['bot']['cycle_time_minutes']
    else:
        cycle_time_minutes = 60
    if 'name' in data['bot']:
        bot_name = data['bot']['name']
    else:
        if gemini_exchange:
            bot_name = "Gemini-" + crypto_currency + "-bot"
        else:
            bot_name = "CoinbasePro-" + crypto_currency + "-bot"
    return crypto_currency, buy_amount, dip_percentage,\
        average_period_days, cool_down_period_days,\
        aws_loaded, using_gemini, cycle_time_minutes, bot_name


def get_aws_creds_from_file(config_file: str) -> [str, str, str]:
    """Open a JSON file and get AWS credentials out of it
    Args:
        config_file: Path to the JSON file containing credentials

    Returns:
        aws_access_key: The AWS access key your bot will use
        aws_secret_key: The AWS secret access key
        sns_topic_arn: The SNS topic ARN to publish to
    """
    with open(config_file) as creds_file:
        data = json.load(creds_file)
    aws_access_key = data['aws']['access_key']
    aws_secret_key = data['aws']['secret_access_key']
    sns_topic_arn = data['aws']['sns_arn']
    return aws_access_key, aws_secret_key, sns_topic_arn


def post_to_sns(aws_access_key: str, aws_secret_key: str, sns_topic_arn: str,
                message_subject: str, message_body: str):
    """Post a message and subject to AWS SNS

    Args:
    aws_access_key: The AWS access key your bot will use
    aws_secret_key: The AWS secret access key
    sns_topic_arn: The SNS topic ARN to publish to
    message_subject: A message subject to post to SNS
    message_body: A message body to post to SNS
    """
    sns = boto3.client('sns', region_name="us-east-1",
                       aws_access_key_id=aws_access_key, aws_secret_access_key=aws_secret_key)
    sns.publish(TopicArn=sns_topic_arn, Subject=message_subject, Message=message_body)


def get_average(list_of_numbers: list) -> float:
    """Get an average of the numbers in a list

    Args:
    list_of_numbers: A list of floats

    Returns:
    list_average: A float containing the list average
    """
    sum_of_numbers = 0
    for number in list_of_numbers:
        sum_of_numbers = sum_of_numbers + number
    list_average = sum_of_numbers / len(list_of_numbers)
    return round(list_average, 2)


def dip_percent_value(price: float, percent: float) -> float:
    """Return the value of the current price if it dips a certain percent

    Args:
    price: The price to check a dip percentage against
    percent: the dip percentage we care about

    Returns:
    dip_price: A float containing the price if we hit our dip target
    """
    dip_price = price * (1 - percent / 100)
    return round(dip_price, 2)


def gemini_exchange_cycle(config_file: str, debug_mode: bool) -> None:
    """Perform bot cycles using Gemini as the exchange

        Args:
        config_file: Path to the JSON file containing credentials
        debug_mode: Are we running in debugging mode?
        """
    # Load the configuration file
    config_params = read_bot_config(config_file)
    if config_params[5]:
        aws_config = get_aws_creds_from_file(config_file)
        message = "%s has been started" % config_params[8]
        post_to_sns(aws_config[0], aws_config[1], aws_config[2], message, message)
    # Set API URLs
    if debug_mode:
        gemini_exchange_api_url = "https://api.sandbox.gemini.com"
        mongo_db_connection = "mongodb://bots:buythedip@bots-db:27017/"
    else:
        gemini_exchange_api_url = "https://api.gemini.com"
        mongo_db_connection = "mongodb://bots:buythedip@bots-db:27017/"
    print("LOG: Starting bot...\nLOG: Monitoring %s on Gemini to buy $%s worth"
          " when a %s%% dip occurs." % (config_params[0], config_params[1], config_params[2]))
    print("LOG: Dips are checked against a %s day price"
          " average with a %s day cool down period" % (config_params[3], config_params[4]))
    for cycle in count():
        now = datetime.datetime.now().strftime("%m/%d/%Y-%H:%M:%S")
        print("LOG: Cycle %s: %s" % (cycle, now))
        coin_current_price = gemini_exchange.get_coin_price(
            gemini_exchange_api_url, config_params[0])
        if coin_current_price == -1:
            message = "ERROR: Coin price invalid. This could be an API issue. Ending cycle"
            print(message)
            subject = "Gemini-%s-Coin price invalid" % config_params[0]
            if config_params[5]:
                post_to_sns(aws_config[0], aws_config[1], aws_config[2],
                            subject, message)
            time.sleep(config_params[7] * 60)
            continue
        # Add the current price to the price database
        mongo.add_price(config_params[8], mongo_db_connection, coin_current_price)
        # Verify that there is enough money to transact, otherwise don't bother
        if not gemini_exchange.verify_balance(gemini_exchange_api_url,
                                              config_file, config_params[1]):
            message = "LOG: Not enough account balance" \
                      " to buy $%s worth of %s" % (config_params[1], config_params[0])
            subject = "%s Funding Issue" % config_params[8]
            if config_params[5]:
                post_to_sns(aws_config[0], aws_config[1], aws_config[2],
                            subject, message)
            print("LOG: %s" % message)
            # Sleep for the specified cycle interval then end the cycle
            time.sleep(config_params[7] * 60)
            continue
        # Check if the a week has passed since the last dip buy
        clear_to_proceed = mongo.check_last_buy_date(config_params[8], mongo_db_connection,
                                                     config_params[4])
        if clear_to_proceed is True:
            print("LOG: Last buy date outside cool down period. Checking if a dip is occurring.")
            average_price = mongo.average_pricing(config_params[8], mongo_db_connection,
                                                  config_params[3])
            dip_price = dip_percent_value(average_price, config_params[2])
            print("LOG: A %s%% dip at the average price of %s would be %s"
                  % (config_params[2], average_price, dip_price))
            if coin_current_price <= dip_price:
                print("LOG: The current price of %s is <= %s. We are in a dip!"
                      % (coin_current_price, dip_price))
                did_buy = gemini_exchange.buy_currency(gemini_exchange_api_url,
                                                       config_file,
                                                       config_params[0], config_params[1])
                message = "Buy success status is %s for %s worth of %s" \
                          % (did_buy, config_params[1], config_params[0])
                subject = "%s Buy Status Alert" % config_params[8]
                mongo.set_last_buy_date(config_params[8], mongo_db_connection)
                print("LOG: %s" % message)
                if config_params[5]:
                    post_to_sns(aws_config[0], aws_config[1], aws_config[2],
                                subject, message)
            else:
                print("LOG: The current price of %s is > %s. We are not in a dip!"
                      % (coin_current_price, dip_price))
        else:
            print("LOG: Last buy date inside cool down period. No buys will be attempted.")

        # Run a price history cleanup daily otherwise sleep the interval
        if (cycle * config_params[7]) % 1440 == 0:
            print("LOG: Cleaning up price history older than 30 days.")
            mongo.cleanup_old_records(config_params[8], mongo_db_connection)
        else:
            # Sleep for the specified cycle interval
            time.sleep(config_params[7] * 60)


def coinbase_pro_cycle(config_file: str, debug_mode: bool) -> None:
    """Perform bot cycles using Coinbase Pro as the exchange

        Args:
        config_file: Path to the JSON file containing credentials
        debug_mode: Are we running in debugging mode?
        """
    # Load the configuration file
    config_params = read_bot_config(config_file)
    if config_params[5]:
        aws_config = get_aws_creds_from_file(config_file)
        message = "%s has been started" % config_params[8]
        post_to_sns(aws_config[0], aws_config[1], aws_config[2], message, message)
    # Set API URLs
    if debug_mode:
        coinbase_pro_api_url = "https://api-public.sandbox.pro.coinbase.com/"
        mongo_db_connection = "mongodb://bots:buythedip@bots-db:27017/"
    else:
        coinbase_pro_api_url = "https://api.pro.coinbase.com/"
        mongo_db_connection = "mongodb://bots:buythedip@bots-db:27017/"
    print("LOG: Starting bot...\nLOG: Monitoring %s on Coinbase Pro to buy $%s worth"
          " when a %s%% dip occurs." % (config_params[0], config_params[1], config_params[2]))
    print("LOG: Dips are checked against a %s day price"
          " average with a %s day cool down period" % (config_params[3], config_params[4]))
    for cycle in count():
        now = datetime.datetime.now().strftime("%m/%d/%Y-%H:%M:%S")
        print("LOG: Cycle %s: %s" % (cycle, now))
        coin_current_price = coinbase_pro.get_coin_price\
            (coinbase_pro_api_url, config_file, config_params[0])
        # Add the current price to the price database
        mongo.add_price(config_params[8], mongo_db_connection, coin_current_price)
        # Verify that there is enough money to transact, otherwise don't bother
        if not coinbase_pro.verify_balance(coinbase_pro_api_url, config_file, config_params[1]):
            message = "LOG: Not enough account balance" \
                      " to buy $%s worth of %s" % (config_params[1], config_params[0])
            subject = "%s Funding Issue" % config_params[8]
            if config_params[5]:
                post_to_sns(aws_config[0], aws_config[1], aws_config[2],
                            subject, message)
            print("LOG: %s" % message)
            # Sleep for the specified cycle interval then end the cycle
            time.sleep(config_params[7] * 60)
            continue
        # Check if the a week has passed since the last dip buy
        clear_to_proceed = mongo.check_last_buy_date(config_params[8], mongo_db_connection,
                                                     config_params[4])
        if clear_to_proceed is True:
            print("LOG: Last buy date outside cool down period. Checking if a dip is occurring.")
            average_price = mongo.average_pricing(config_params[8], mongo_db_connection,
                                                  config_params[3])
            dip_price = dip_percent_value(average_price, config_params[2])
            print("LOG: A %s%% dip at the average price of %s would be %s"
                  %(config_params[2], average_price, dip_price))
            if coin_current_price <= dip_price:
                print("LOG: The current price of %s is <= %s. We are in a dip!"
                      % (coin_current_price, dip_price))
                did_buy = coinbase_pro.buy_currency(coinbase_pro_api_url,
                                                    config_file, config_params[0], config_params[1])
                message = "Buy success status is %s for %s worth of %s"\
                          % (did_buy, config_params[1], config_params[0])
                subject = "%s Buy Status Alert" % config_params[8]
                mongo.set_last_buy_date(config_params[8], mongo_db_connection)
                print("LOG: %s" % message)
                if config_params[5]:
                    post_to_sns(aws_config[0], aws_config[1], aws_config[2],
                                subject, message)
            else:
                print("LOG: The current price of %s is > %s. We are not in a dip!"
                      % (coin_current_price, dip_price))
        else:
            print("LOG: Last buy date inside cool down period. No buys will be attempted.")

        # Run a price history cleanup daily otherwise sleep the interval
        if (cycle * config_params[7]) % 1440 == 0:
            print("LOG: Cleaning up price history older than 30 days.")
            mongo.cleanup_old_records(config_params[8], mongo_db_connection)
        else:
            # Sleep for the specified cycle interval
            time.sleep(config_params[7] * 60)
